DXIL: Use correct type ID when writing ValueAsMetadata.

maleadt opened the pull request:
Small drive-by PR (I'm not working on DXIL, but am working on a similar IR downgrader based on the opaque pointer handling developed for DXIL): It looks like the `ValueAsMetadata` handling currently is broken when writing references to functions. For example:

```llvm
target datalayout = "e-m:e-p:32:32-i1:32-i8:8-i16:16-i32:32-i64:64-f16:16-f32:32-f64:64-n8:16:32:64"
target triple = "dxil-unknown-shadermodel6.7-library"

define void @kernel(ptr addrspace(1)) {
    ret void
}

!kernel = !{!0}
!0 = !{ptr @kernel}
```

Assembling this with `llc` and disassembling the contained IR with `llvm-dis` from LLVM 12 (again, I'm not working on DXIL so I don't have `dxil-dis` handy) breaks:

```
Assertion failed: (V && "Unexpected null Value"), function get, file Metadata.cpp, line 350.
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /Users/tim/Julia/src/llvm-12/build/bin/llvm-dis reduced.dxil -o -
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  llvm-dis                 0x0000000102f421bc llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 80
1  llvm-dis                 0x0000000102f42720 PrintStackTraceSignalHandler(void*) + 28
2  llvm-dis                 0x0000000102f40890 llvm::sys::RunSignalHandlers() + 140
3  llvm-dis                 0x0000000102f43c14 SignalHandler(int) + 276
4  libsystem_platform.dylib 0x000000018611f584 _sigtramp + 56
5  libsystem_pthread.dylib  0x00000001860eec20 pthread_kill + 288
6  libsystem_c.dylib        0x0000000185ffba30 abort + 180
7  libsystem_c.dylib        0x0000000185ffad20 err + 0
8  llvm-dis                 0x0000000102ce8db4 llvm::ValueAsMetadata::get(llvm::Value*) + 100
9  llvm-dis                 0x00000001029d672c llvm::MetadataLoader::MetadataLoaderImpl::parseOneMetadata(llvm::SmallVectorImpl<unsigned long long>&, unsigned int, (anonymous namespace)::(anonymous namespace)::PlaceholderQueue&, llvm::StringRef, unsigned int&) + 2184
10 llvm-dis                 0x00000001029d5a78 llvm::MetadataLoader::MetadataLoaderImpl::parseMetadata(bool) + 1448
11 llvm-dis                 0x00000001029df7fc llvm::MetadataLoader::parseMetadata(bool) + 60
12 llvm-dis                 0x000000010298634c llvm::MetadataLoader::parseModuleMetadata() + 40
13 llvm-dis                 0x0000000102982f2c (anonymous namespace)::BitcodeReader::parseModule(unsigned long long, bool, llvm::function_ref<llvm::Optional<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > (llvm::StringRef)>) + 2272
```

IIUC, this is because the value emitted here is of type `typedptr(void (ptr addrspace(1)), 0)`, while the type is simply `void (i8*)*` because it's not looked up in the PointerAnalysisMap. Passing `V` along fixes this.

cc @bogner @llvm-beanz

maleadt committed:
DXIL: Use correct type ID when writing ValueAsMetadata.

maleadt commented:
A similar case for loading pointer values:

```llvm
define void @kernel(ptr %0) {
  %box = alloca ptr addrspace(1)
  %x = load ptr addrspace(1), ptr %box, align 8
  ret void
}
```

This currently throws a `Mismatch in load/store type: expected i8* but found i8 addrspace(1)*` when reading the bitcode because the `load` type is set to `i8*`, while `i8 addrspace(1)*` is expected. Passing the instruction along similarly to the change proposed here makes this work:

```patch
diff --git a/llvm/lib/Bitcode/Writer50/BitcodeWriter50.cpp b/llvm/lib/Bitcode/Writer50/BitcodeWriter50.cpp
index c5c52932512f..20b58f7d1846 100644
--- a/llvm/lib/Bitcode/Writer50/BitcodeWriter50.cpp
+++ b/llvm/lib/Bitcode/Writer50/BitcodeWriter50.cpp
@@ -2845,7 +2845,7 @@ void ModuleBitcodeWriter50::writeInstruction(const Instruction &I,
       if (!pushValueAndType(I.getOperand(0), InstID, Vals)) // ptr
         AbbrevToUse = FUNCTION_INST_LOAD_ABBREV;
     }
-    Vals.push_back(getTypeID(I.getType()));
+    Vals.push_back(getTypeID(I.getType(), &I));
     Vals.push_back(getEncodedAlign(cast<LoadInst>(I).getAlign()));
     Vals.push_back(cast<LoadInst>(I).isVolatile());
     if (cast<LoadInst>(I).isAtomic()) {
```

bogner reviewed (commented):
Pointer values are a bit interesting because DXIL (and its usual source language, HLSL) don't generally expose pointers, so these code paths are probably not really reachable without fiddling with the IR we input to the DXIL backend. That said, I think it's reasonable to take the obvious fixes either way.

It would be better if this came with a test. Can you come up with something sensible or do you need someone (like me) with access to dxil-dis to take a stab at that?

llvm-beanz commented:
Unfortunately the always casting to and from the typed pointer doesn't work for DXIL because DXIL disallows bitcasts for a lot of cases, and that is enforced by the DXIL validator which operates on typed pointer IR.

maleadt commented:
Sorry for the delay. Added a test (verifying that the test crashes without this change), and also changed the `dxil-dis` CMakeLists.txt to checkout the `main` branch from https://github.com/microsoft/DirectXShaderCompiler instead of the `master` branch my CMake defaults to.

maleadt commented:
@bogner Anything else you want me to add here?

bogner reviewed (approved):
LGTM. Thanks!

maleadt commented:
Feel free to merge, I don't have commit bit.
